### PR TITLE
Load an ANSTO EMU datafile (no ticket)

### PR DESCRIPTION
**Pull request by geishm-ansto**
Loads an ASNTO EMU datafile as a TOF EventWorkspace.

No issue number allocated - I am an external contributor working for ANSTO.

Contains a unit test LoadEMUauTest.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by geishm-ansto** [comment]
Applied the clang-format patch to the master and pushed back up.

**Comment by geishm-ansto** [comment]
Don't understand why cppcheck is failing still - the errors were fixed with the update eg _cancel is still flagged as an unused variable but it was removed in the update.

**Comment by martyngigg** [comment]
@geishm-ansto It looks like the failures are due to similar fields in the`.cpp` file and your fixes above only changed things in the `.h` file.

**Comment by geishm-ansto** [comment]
Apologies, there was a remnant LoadANSTOEventFile.cpp file that was removed from the repository. The clang-format check is using the latest revision 'b7bd301..' but  the cpp-check refers to rev '364f1..' and is failing on the file that was deleted.

**Comment by geishm-ansto** [comment]
@martyngigg I can't determine the segmentation fault on the Linux platforms from the error messages. I think my only option is to clone the code on a local Linux machine to solve the problem. Unless you have a suggestion, I should close this pull request and come back later. Thanks for your patience.

**Comment by martyngigg** [comment]
@geishm-ansto The segfault appears to occur when [retrieving instrument parameters](https://github.com/mantidproject/mantid/pull/23713/files#diff-dd771b9d041331dca8b354c9fe55fadbR594). It looks like the instrument parameter file has failed to be found on Linux because we extract the instrument name from the filename of the `_Definition.xml` file and then force it to be upper case. As the filesystem is case sensitive on Linux it fails to find `EMUau_Paramters.xml` as it is looking for `EMUAU_Parameters.xml`. Windows/Mac are not case sensitive so they don't see the same issue.

I have a feeling that this prefix transformation is a hangover from the past as I don't see why it is necessary. I'll check and if others agree then we can just remove [this line](https://github.com/mantidproject/mantid/blob/master/Framework/DataHandling/src/LoadInstrument.cpp#L290) and the test should be good.

**Comment by martyngigg** [comment]
@geishm-ansto I have agreed with others that the current behaviour is odd and we should fix it. Would you be able to remove [this line](https://github.com/mantidproject/mantid/blob/master/Framework/DataHandling/src/LoadInstrument.cpp#L290) to stop the prefix being changed? This should fix your issue.

**Comment by geishm-ansto** [comment]
@martyngigg Thanks, that was the issue - all the checks pass now.

**Comment by geishm-ansto** [comment]
@martyngigg Is there any more I need to do on my end with this pull request?

**Review by soininen — CHANGES_REQUESTED** [review]
The code quality of the new loader looks superb, although I couldn't test this myself as no test instructions nor data was provided. Hopefully the unit test is sufficient. I couldn't find any changes to the release notes, so please consider adding a line to `docs/sources/release/v3.14.0/indirect_inelastic.rst` or whatever is suitable.

That said there were a few minor issues in the code, please see the comments below. Additionally, I would've hoped to see more `const` keywords peppered around the function parameters and local variables.

File: Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+  DECODE_OOB_BYTE_1,
+  DECODE_OOB_BYTE_2,
+  // for all events
+  DECODE_VAL_BITFIELDS,
+  DECODE_DT // final state - then output data and return to DECODE_START
+};
+
+/*
+
+// Types of OOB events, and 'NEUTRON' event.  Not all are used for all
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h]: Cleaned up the comment formatting.

File: Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+  // note: in the old format 0x00010001, the evt_stg_nbits_wa did not exist and
+  // it contained evt_stg_xy_signed
+  if (hdr_base.format_number <= 0x00010001) {
+    hdr_packed.evt_stg_xy_signed = hdr_packed.evt_stg_nbits_wa;
+    hdr_packed.evt_stg_nbits_wa = 0;
+  }
+
+  // Setup the clock_scale.  In format 0x00010001 this was not part of the
+  // headers, hence a command line option -c is provided to allow it to be
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h]: Changed to function argument.

File: Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+  // specified manually. In the current format 0x00010002, clock_scale is
+  // written to the header and need not be specified, unless some alternate
+  // scale is needed.
+  double scale_microsec = hdr_base.clock_scale / 1000.0;
+  if (!hdr_base
+           .clock_scale) { // old eventfile format did not have clock_scale...
+    scale_microsec = def_clock_scale / 1000.0;
+    ;
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h]: Removed.

File: Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+  }
+
+  // Setup the clock_scale.  In format 0x00010001 this was not part of the
+  // headers, hence a command line option -c is provided to allow it to be
+  // specified manually. In the current format 0x00010002, clock_scale is
+  // written to the header and need not be specified, unless some alternate
+  // scale is needed.
+  double scale_microsec = hdr_base.clock_scale / 1000.0;
+  if (!hdr_base
+           .clock_scale) { // old eventfile format did not have clock_scale...
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h]: Done.

File: Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+"UNKNOWN",
+// RAW = -8, BEAM_MONITOR = -7, VETO = -6, FRAME_DEASSERT = -5, FLUSH = -4,
+FRAME_AUX_START = -3, FRAME_START = -2, DATASIZES = -1 "RAW", "BEAM_MONITOR",
+"VETO", "FRAME_DEASSERT", "FLUSH", "FRAME_AUX_START", "FRAME_START", "DATASIZES"
+};
+
+*/
+
+template <class IReader, class IEventHandler, class IProgress>
+void ReadEventFile(IReader &loader, IEventHandler &handler, IProgress &progress,
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h]: I am not the original author of this code but appreciate the complexity of decoding of an existing intricate binary packing format.  I agree the state machine could be cleaned up but the intent was to re-use a working decoder with a wrapper to make it useable. I would be reluctant to make structural changes now.

File: Framework/DataHandling/inc/MantidDataHandling/LoadEMU.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+#include "MantidAPI/IFileLoader.h"
+#include "MantidAPI/LogManager.h"
+#include "MantidDataObjects/EventWorkspace.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidNexus/NexusClasses.h"
+
+namespace Mantid {
+namespace DataHandling {
+/*
+Loads an Emu data file. Implements API::IFileLoader and its file check methods
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/inc/MantidDataHandling/LoadEMU.h]: I couldn't find the reference to copyright on the mantid documentation nor the python update script. I updated the header based on the LoadISISNexus2.h format and added the required and optional parameters.

File: Framework/DataHandling/inc/MantidDataHandling/LoadEMU.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+File change history is stored at: <https://github.com/mantidproject/mantid>
+Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+
+class DLLExport LoadEMU : public API::IFileLoader<Kernel::FileDescriptor> {
+
+  DataObjects::EventWorkspace_sptr m_localWorkspace;
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/inc/MantidDataHandling/LoadEMU.h]: Done.

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 8]
```diff
@@ -0,0 +1,931 @@
+#include <math.h>
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Moved and dropped <stdio.h>

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+#include <Poco/Util/PropertyFileConfiguration.h>
+
+namespace Mantid {
+namespace DataHandling {
+
+using namespace Kernel;
+using ANSTO::EventVector_pt;
+
+// Detector consts
+static const size_t DETECTORS = 51;
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Agreed and changed.

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 10]
```diff
@@ -0,0 +1,931 @@
...
+
+using TimeLimits = std::pair<double, double>;
+
+namespace EMU {
+
+// map the comma separated range of indexes to the vector via a lambda function
+// throws an exception if it is outside the vector range
+//
+template <typename T, typename F>
+void mapRangeToIndex(const std::string &line, std::vector<T> &result,
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Moved the local functions to the anonymous namespace but kept EMU to capture the changes from the existing class definitions.

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 11]
```diff
@@ -0,0 +1,931 @@
...
+    return tof * 1.0e6;
+  }
+};
+
+// Implement emu specific handlers for the more general EMU events. The main
+// differences with the current impementation of handlers:
+// - tof needs to be derived from the observed time because of the doppler drive
+// - emu includes direct and indirect virtual detectors that account for
+// alternate paths
+// - the loader returns returns the doppler and auxillary time along with the
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Fixed.

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 12]
```diff
@@ -0,0 +1,931 @@
...
+
+  // mask
+  exts.clear();
+  exts.emplace_back(".xml");
+  declareProperty(Kernel::make_unique<API::FileProperty>(
+                      MaskStr, "", API::FileProperty::OptionalLoad, exts),
+                  "The input filename of the mask data");
+
+  declareProperty(
+      Kernel::make_unique<Kernel::PropertyWithValue<std::string>>(
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Done.

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 13]
```diff
@@ -0,0 +1,931 @@
...
+ * Creates an event workspace and sets the title.
+ */
+void LoadEMU::createWorkspace(ANSTO::Tar::File &tarFile) {
+
+  // Clean up and check file
+  // -----------------------
+  // Delete the output workspace name if it existed
+  std::string outName = getPropertyValue("OutputWorkspace");
+  if (API::AnalysisDataService::Instance().doesExist(outName))
+    API::AnalysisDataService::Instance().remove(outName);
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: I am still not fluent in the framework and copied existing code. Have removed the calls.

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 14]
```diff
@@ -0,0 +1,931 @@
...
+
+/**
+ * Execute the algorithm. The steps involved are:
+ *   Get the instrument properties and load options
+ *   Create the workspace
+ *   Load the instrument from the IDF
+ *   Reposition the relevant neutronic values for model based on the parameters
+ *   Load the data values and convert to TOF
+ */
+void LoadEMU::exec() {
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Extracted several smaller functions to improve the readability.

File: docs/source/algorithms/LoadEMU-v1.rst [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+.. relatedalgorithms::
+
+.. properties::
+
+Description
+-----------
+
+Load data from the EMU beamline at ANSTO. The workspace generated is a TOF EventWorkspace.
+
+.. categories::
```
[the remark this answers is not in the document]
Comment (geishm-ansto) [reply docs/source/algorithms/LoadEMU-v1.rst]: Moved to the end.

File: Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/DataHandling/inc/MantidDataHandling/LoadANSTOEventFile.h]: I understand and agree. Better leave it as-is.

**Comment by soininen** [comment]
@geishm-ansto Looks like you have to `#include<exception>` or something similar.

File: Framework/DataHandling/inc/MantidDataHandling/LoadEMU.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/DataHandling/inc/MantidDataHandling/LoadEMU.h]: I think the copyright changed after you branched out from master. Please, merge master, or check how the copyright looks there.

[the remark this answers is not in the document]
Comment (soininen) [reply Framework/DataHandling/inc/MantidDataHandling/LoadEMU.h]: Also the script mentioned above is available in the latest master.

**Comment by geishm-ansto** [comment]
@soininen I've fixed those exceptions. My development is on Windows but I can't find any options on Visual Studio that ignore Windows deviations from the standards. It would make my life easier. :-)

**Comment by soininen** [comment]
@geishm-ansto Every compiler seems to have its own little quirks. I'm on gcc and have to occasionally fix stuff that Visual Studio's better diagnostics finds. Anyway, this looks good to me now. Thanks a lot for the high quality contribution!

**Review by soininen — APPROVED** [review]

**Pull request merged**